---
title: "Figure4 and Supp Figure6"
author: "Huaying Qiu"
date: "2024-12-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

This Rmd include the code to reproduce the plots related to proteomics data in figure 4. <ins>**Note that for most chunks `eval = FALSE`**</ins>. If one wish to knit the Rmd, instead of running the chunks one by one, one should set `eval = TRUE`.

# Packages

```{r, message=FALSE, warning=FALSE}

library(tidyverse)
library(ggpubr)
library(ComplexHeatmap)

```

# Load and Process Data

## Annotation

Mainly three columns from these dataset were used:

- `coreName`: Identifier of tissue sample.
- `cellLabel`: Unique cell ID generated by the segmentation.
- `Annotation`: Cell type annotation.  

These data frames served as the bases which the functional markers were later joined to. 

```{r, message=FALSE, warning=FALSE, eval=FALSE}

INDEPTH_annotation <- read_csv('~/INDEPTH/run_041624/output/Annotation/final_annotation_edge_corrected.csv')


INDEPTH_ROI <- read_csv('~/INDEPTH/run_041624/output/Annotation/ROI_with_full_marker_no_0_syto13.csv')

```

## Color palette for cell types

Define color palettes for different cell types. 

```{r}

celltype_color <- c(Tumor = "#e51d1d", 
                    `CD8 T` = "#377eb8", 
                    `CD4 T` = "#4eaf49", 
                    `M1-like` = "#974da2", 
                    `M2-like` = "#ff8000", 
                    `DC` = "#83e3f0", 
                    Endothelial = "#c67f66", 
                    Treg = "#9ED900", 
                    Other = 'gray',
                    `B cell` = "#e89ff0")

tumor_color <- c(`Other Tumor` = "#B22222", 
                 `Tumor BCL2` = "#008080",
                 `Tumor Myc` =  "#DAA520",
                 `Tumor BCL6` = "#4682B4",
                 Tumor = "#e51d1d")

```

## Selected cores

`coreName` for cores that were included in the analysis. 

```{r}

selected_cores <- c("Rochester_4", "Rochester_6",
                    "Rochester_7", "Rochester_9", "Rochester_11", "Rochester_12",
                    "Rochester_13", "Rochester_14",
                    "Rochester_15", "Rochester_16", "Rochester_17", "Rochester_18",
                    "Rochester_19", "Rochester_21", "Rochester_23", 'Rochester_25',
                    "Rochester_TonsilA", "DFCI_2.2", "DFCI_3.2",
                    "DFCI_4.1", "DFCI_7.1", "DFCI_8.1",
                    "DFCI_12.1", "DFCI_13.2", "DFCI_14.1", "DFCI_15.2", "DFCI_17.1",
                    "DFCI_18.2", "DFCI_19.2", "DFCI_22.2", "DFCI_23.2", "DFCI_Tonsil1")

```

## Load functional marker captured on Fusion



```{r, eval = FALSE}

# List the path of all the csvs for all cores.

Rochester_path <- list.files('~/INDEPTH/run_041624/output/extracted_info_16bit_sub_cycle15_processed_gated/Rochester', 'dataScaleSize.csv', full.names = TRUE, recursive = TRUE)

DFCI_path <- list.files('~/INDEPTH/run_041624/output/extracted_info_16bit_sub_cycle15_processed_gated/DFCI', 'dataScaleSize.csv', full.names = TRUE, recursive = TRUE)

# Read those csvs into a list

Rochester_df_list <- map(as.list(Rochester_path), function(x){
  df <- read_csv(x) %>% 
    mutate(coreName = paste0('Rochester_',as.character(str_extract(x, '(?<=Rochester/).*(?=/)'))))
  return(df)
})

DFCI_df_list <- map(as.list(DFCI_path), function(x){
  df <- read_csv(x) %>% 
    mutate(coreName = paste0('DFCI_', as.character(str_extract(x, '(?<=DFCI/).*(?=/)'))))
  return(df)
})

# Filter out csvs with 0 rows just in case, but there shouldn't be any

DFCI_df <- Filter(function(x) nrow(x) > 0, DFCI_df_list)
Rochester_df <- Filter(function(x) nrow(x) > 0, Rochester_df_list)

# rbind all the data frames in the list

DFCI_df <- do.call(rbind, DFCI_df)
Rochester_df <- do.call(rbind, Rochester_df)

# full_join the data from the two cohorts 

df_fusion <- full_join(DFCI_df, Rochester_df) %>% 
  filter(coreName %in% selected_cores)

```

## Load functional marker captured on GeoMx

```{r, eval = FALSE}

# List the path of all the csvs for all cores.

Rochester_path_geomx <- list.files('~/INDEPTH/run_041624/output/extracted_info_geomx_gated_121724/Rochester', 'dataScaleSize.csv', full.names = TRUE, recursive = TRUE)
DFCI_path_geomx <- list.files('~/INDEPTH/run_041624/output/extracted_info_geomx_gated_121724/DFCI', 'dataScaleSize.csv', full.names = TRUE, recursive = TRUE)

# Read those csvs into a list

Rochester_geomx_df_list <- map(as.list(Rochester_path_geomx), function(x){
  df <- read_csv(x) %>% 
    mutate(coreName = paste0('Rochester_', as.character(str_extract(x, '(?<=Rochester/).*(?=/)'))))
  return(df)
})

DFCI_geomx_df_list <- map(as.list(DFCI_path_geomx), function(x){
  df <- read_csv(x) %>% 
    mutate(coreName = paste0('DFCI_', as.character(str_extract(x, '(?<=DFCI/).*(?=/)'))))
  return(df)
})

# Filter out csvs with 0 rows just in case, but there shouldn't be any

DFCI_df_geomx <- Filter(function(x) nrow(x) > 0, DFCI_geomx_df_list)
Rochester_df_geomx <- Filter(function(x) nrow(x) > 0, Rochester_geomx_df_list)

# rbind all the data frames in the list

DFCI_df_geomx <- do.call(rbind, DFCI_df_geomx)
Rochester_df_geomx <- do.call(rbind, Rochester_df_geomx)

# full_join the data from the two cohorts 

df_geomx <- full_join(DFCI_df_geomx, Rochester_df_geomx) %>% 
  filter(coreName %in% selected_cores)

```

## Left join markers to annotation

```{r, eval = FALSE}

df_fusion_with_annotation <- INDEPTH_annotation %>% 
  left_join(df_fusion, by = c('coreName', 'cellLabel'))

# all NAs coming from cores that we didn't select, safe to drop NA 
unique(df_fusion_with_annotation[rowSums(is.na(df_fusion_with_annotation)) > 0, ]$coreName)

df_fusion_with_annotation <- df_fusion_with_annotation %>% 
  drop_na()


```

## Filter out cells that were artifacts

Upon visual inspection of some of the segmentation masks, we found that some of the segmentation masks include imaging artifacts (out of focus areas, blobs with abnormally high pixel intensity, etc.). Therefore, we generated binary masks in imageJ to filter out those cells. Two segments with abnormally high CD45RO was also removed. They were visually inspected to ensure that they were artifacts rather than actual cells.

```{r, eval = FALSE}

cell_filter_files <- list.files('~/INDEPTH/run_041624/output/cells_to_filter', full.names = TRUE, recursive = TRUE)

cell_filter_df_list <- map(as.list(cell_filter_files), function(x){
  df <- read_table(x, col_names = c('cellLabel')) %>% 
    mutate(coreName = str_extract(x, '(?<=cells_to_filter/).*(?=.txt)'))
})

cell_filter_df <- do.call(rbind, cell_filter_df_list)


sum(is.na(cell_filter_df))


df_fusion_with_annotation_filtered <- df_fusion_with_annotation %>% 
  mutate(coreName_cellLabel = paste0(coreName, '_', cellLabel)) %>% 
  filter(!(coreName_cellLabel %in% paste0(cell_filter_df$coreName, '_', cell_filter_df$cellLabel)))

df_geomx_filtered <- df_geomx %>% 
  mutate(coreName_cellLabel = paste0(coreName, '_', cellLabel)) %>% 
  filter(!(coreName_cellLabel %in% paste0(cell_filter_df$coreName, '_', cell_filter_df$cellLabel)))


CD45RO_outlier_core_cellLabel <- df_fusion_with_annotation_filtered %>% 
  filter(CD45RO > 2800) %>% 
  dplyr::select(coreName, cellLabel) %>% 
  mutate(coreName_cellLabel = paste0(coreName, '_', cellLabel))

df_fusion_with_annotation_filtered <- df_fusion_with_annotation_filtered %>% 
  filter(!(coreName_cellLabel %in% CD45RO_outlier_core_cellLabel$coreName_cellLabel))

df_geomx_filtered <- df_geomx_filtered %>% 
  filter(!(coreName_cellLabel %in% CD45RO_outlier_core_cellLabel$coreName_cellLabel))

```

## Join Fusion and GeoMx data set

Tonsils were not included in the downstream analysis. They were there to serve as technical controls. 

```{r, eval = FALSE}

INDEPTH_whole_core_df <- df_geomx_filtered %>% 
  left_join(df_fusion_with_annotation_filtered, by = c('coreName', 'cellLabel')) %>% 
  filter(!str_detect(coreName, 'Tonsil'))

```

## Median nuclear signal scaling 

Median nuclear signal of each core was used to compensate for the different binding efficiency of cores. Since the these functional markers were extracted from different intermediate qptiffs generated by the Fusion and each qptiff had its own DAPI channel, markers were scaled according to their corresponding DAPI. 

```{r, eval = FALSE}

markers <- c('CD45RO', 'CD45RA', 'Ki67', 'HLA1', 'HLA-DR', 'PD-1', 'LAG3', 'GZMB', 'Tox', 'PD-L1')

INDEPTH_whole_core_df_norm <- INDEPTH_whole_core_df


mapping <- list(
  DAPI_cycle4 = c('CD45RO', 'CD45RA'),
  # DAPI_cycle6 = c('Cdt1', 'Geminin'),
  DAPI_cycle7 = c('Ki67'),
  DAPI_cycle9 = c('HLA1', 'HLA-DR'),
  DAPI_cycle10 = c('PD-1','LAG3'),
  DAPI_cycle11 = c('GZMB'),
  Syto13 = c('Tox', 'PD-L1')
)

for (dapi_cycle in names(mapping)) {
  markers <- mapping[[dapi_cycle]]
  INDEPTH_whole_core_df_norm <- INDEPTH_whole_core_df %>%
    group_by(coreName) %>%
    mutate(across(all_of(markers), ~ .x / median(.data[[dapi_cycle]])))
}

INDEPTH_whole_core_df_norm <- ungroup(INDEPTH_whole_core_df_norm)

```

## Scale to [0,1]

Markers were then scaled to [0,1] so that they can be compared across cores. 

```{r, eval = FALSE}

markers <- c('CD45RO', 'CD45RA', 'Ki67', 'HLA1', 'HLA-DR', 'PD-1', 'LAG3', 'GZMB', 'Tox', 'PD-L1')

INDEPTH_whole_core_df_scaled_norm <- INDEPTH_whole_core_df_norm %>% 
  mutate(across(all_of(markers), ~(.x - min(.x))/(max(.x) - min(.x)))) %>%
  # mutate(across(all_of(marker_vec), ~(.x - min(.x))/(max(.x) - min(.x)))) %>%
  mutate(across(all_of(markers), ~ifelse(.x < 0, 0, .x))) %>%
  mutate(across(all_of(markers), ~ifelse(.x > 1, 1, .x))) %>% 
  rename(X_cent_geomx = 'X_cent',
         Y_cent_geomx = 'Y_cent',
         X_cent_fusion = 'X_cent.x',
         Y_cent_fusion = 'Y_cent.x') %>% 
  dplyr::select(-cellSize.x, -coreName_cellLabel.x, -Y_cent.y, -X_cent.y, - coreName_cellLabel.y)

INDEPTH_ROI_df_scaled_norm <- INDEPTH_ROI %>% 
  dplyr::select(cellLabel, coreName) %>% 
  left_join(INDEPTH_whole_core_df_scaled_norm, by = c('cellLabel', 'coreName')) %>% 
  drop_na()

write_csv(INDEPTH_whole_core_df_scaled_norm, '~/INDEPTH/run_041624/output/Scaled_data_121724/wholeCore.csv')

write_csv(INDEPTH_ROI_df_scaled_norm, '~/INDEPTH/run_041624/output/Scaled_data_121724/ROI.csv')

```

## Load phenotyping markers

```{r, eval = FALSE}

annotation_with_pheno_marker <- read_csv('~/INDEPTH/run_041624/output/Annotation/annotation_with_full_marker.csv')

pheno_marker <- c("CD20", "Pax5", "CD3", "CD8", "CD4", "FoxP3", "CD11c", "CD15", "CD31", "CD68", "CD163", "DAPI")

annotation_with_pheno_marker <- annotation_with_pheno_marker %>% 
  dplyr::select(cellLabel, coreName, all_of(pheno_marker), Annotation)

annotation_with_pheno_marker <- INDEPTH_whole_core_df_scaled_norm %>% 
  dplyr::select(cellLabel, coreName) %>% 
  left_join(annotation_with_pheno_marker, by = c('cellLabel', 'coreName')) %>% 
  drop_na()


annotation_with_pheno_marker <- annotation_with_pheno_marker %>%
  group_by(coreName) %>% 
  mutate(across(all_of(pheno_marker), ~./median(DAPI))) %>% 
  ungroup() %>% 
  dplyr::select(-DAPI)

```
## Load Motif information

```{r, eval = FALSE}

niche_df <- read_csv('~/INDEPTH/run_041624/output/hop_analysis_121724/cellLabel_coreName_niche_one_hop_count.csv') %>% 
  mutate(Niche = factor(Niche)) %>%  
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative'))

```

# Figure plots

## Fig4 C


```{r, eval =FALSE}

pheno_marker <- c("CD20", "Pax5", "CD3", "CD8", "CD4", "FoxP3", "CD11c", "CD15", "CD31", "CD68", "CD163")

heatmap_df <- annotation_with_pheno_marker %>% 
  ungroup() %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  #filter(!str_detect(coreName, 'Tonsil')) %>% 
  #filter(str_detect(coreName, 'Rochester')) %>% 
  dplyr::select(Annotation, all_of(pheno_marker))

mean_df <- heatmap_df %>% 
  pivot_longer(cols = -Annotation, names_to = 'marker', values_to = 'value') %>% 
  group_by(marker, Annotation) %>% 
  summarise(mu = mean(value))


pop_mean_df <- heatmap_df %>% 
  pivot_longer(cols = -Annotation, names_to = 'marker') %>% 
  group_by(marker) %>% 
  summarise(pop_mean = mean(value),
            pop_sd = sd(value))

mean_df <- mean_df %>% 
  ungroup() %>% 
  mutate(pop_mean = pop_mean_df$pop_mean[match(mean_df$marker, pop_mean_df$marker)],
         pop_sd = pop_mean_df$pop_sd[match(mean_df$marker, pop_mean_df$marker)]) %>% 
  mutate(z_new = (mu - pop_mean)/pop_sd)

heatmap_df <- mean_df %>% 
  dplyr::select(marker, z_new, Annotation) %>% 
  pivot_wider(names_from = 'marker', values_from = 'z_new')

heatmap_mat <- t(as.matrix(heatmap_df[,2:12]))

colnames(heatmap_mat) <- heatmap_df$Annotation

library(circlize)
col_fun = colorRamp2(c(-2, 0, 2), c('#4575b4', "white", '#d73027'))

bar_vec <- annotation_with_pheno_marker %>% 
  #filter(!str_detect(coreName, 'Tonsil')) %>% 
  #filter(str_detect(coreName, 'Rochester')) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  group_by(Annotation) %>% 
  dplyr::count() %>% 
  mutate(log_n = log10(n))

column_ha <- HeatmapAnnotation(count = anno_barplot(bar_vec$n, border = FALSE, bar_width = 0.9,
                                                    gp = gpar(col = 'black', fill = rep(celltype_color[sort(names(celltype_color))], 2)),
                                                    height = unit(3, 'cm')),
                               show_annotation_name = FALSE)

column_ha2 <- HeatmapAnnotation(celltype = sort(names(celltype_color)),
                                col = list(celltype = celltype_color), show_annotation_name = TRUE,
                                gp = gpar(col = 'black', lwd = 2),
                                annotation_legend_param = list(title = 'Cell Type'))

fig4_c <- Heatmap(heatmap_mat, cluster_rows = FALSE, cluster_columns  = FALSE, col = col_fun, clustering_method_columns = 'complete',
                   row_names_side = 'left', top_annotation = column_ha, column_names_gp = grid::gpar(fontsize = 12),
                   row_names_gp = grid::gpar(fontsize = 12), rect_gp = gpar(col = "black", lwd = 2), show_column_names = TRUE, name = 'Z Score',
                   heatmap_legend_param = list(border = 'black'))


draw(fig4_c, ht_gap = unit(0.1, 'cm'))



```

## Fig4 D

```{r, eval = FALSE}

p1 <- annotation_with_pheno_marker %>% 
  filter(Annotation != 'Other') %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  filter(Annotation != 'Neutrophil') %>% 
  mutate(Annotation = factor(Annotation, levels = c('Tumor', 'CD4 T', 'CD8 T', 'Treg', 'DC', 'M1-like', 'M2-like', 'Endothelial'))) %>% 
  filter(ebv_status == 'Positive') %>% 
  group_by(coreName, Annotation) %>% 
  count() %>% 
  group_by(coreName) %>% 
  mutate(order_vec = n[Annotation == 'Tumor']/sum(n)) %>% 
  ungroup() %>% 
  ggplot(aes(x = reorder(coreName, -order_vec), y = n, fill = Annotation)) + 
  geom_col(position = 'fill') + 
  geom_hline(yintercept = 0.5, linetype = 'dashed') + 
  scale_fill_manual(values = celltype_color, breaks = c('Tumor', 'CD4 T', 'CD8 T', 'Treg', 'DC', 'M1-like', 'M2-like', 'Endothelial')) + 
  #scale_fill_brewer(palette = 'Set1') + 
  labs(x = 'Core Name', y = 'Proportion', fill = 'Cell Type', title = 'Cell Composition of EBV+ Cores') + 
  theme_bw() + 
  theme(axis.text.x = element_text(face = 'bold', angle = 90),
        axis.text.y = element_text(face = 'bold'),
        axis.title = element_text(face = 'bold'),
        title = element_text(face = 'bold'))


p2 <- annotation_with_pheno_marker %>% 
  filter(Annotation != 'Other') %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  filter(Annotation != 'Neutrophil') %>% 
  mutate(Annotation = factor(Annotation, levels = c('Tumor', 'CD4 T', 'CD8 T', 'Treg', 'DC', 'M1-like', 'M2-like', 'Endothelial'))) %>% 
  filter(ebv_status == 'Negative') %>% 
  group_by(coreName, Annotation) %>% 
  count() %>% 
  group_by(coreName) %>% 
  mutate(order_vec = n[Annotation == 'Tumor']/sum(n)) %>% 
  ungroup() %>% 
  ggplot(aes(x = reorder(coreName, -order_vec), y = n, fill = Annotation)) + 
  geom_col(position = 'fill') + 
  geom_hline(yintercept = 0.5, linetype = 'dashed') + 
  scale_fill_manual(values = celltype_color, breaks = c('Tumor', 'CD4 T', 'CD8 T', 'Treg', 'DC', 'M1-like', 'M2-like', 'Endothelial')) + 
  #scale_fill_brewer(palette = 'Set1') + 
  labs(x = 'Core Name', y = 'Proportion', fill = 'Cell Type', title = 'Cell Composition of EBV+ Cores') + 
  theme_bw() + 
  theme(axis.text.x = element_text(face = 'bold', angle = 90),
        axis.text.y = element_text(face = 'bold'),
        axis.title = element_text(face = 'bold'),
        title = element_text(face = 'bold'))

p1

p2

```



## Fig4 E

```{r, eval = FALSE}

log2fc_p <- annotation_with_pheno_marker %>% 
  filter(!str_detect(coreName, 'Tonsil')) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  filter(Annotation != 'Neutrophil') %>% 
  filter(Annotation != 'Tumor') %>%
  filter(Annotation != 'Other') %>% 
  filter(Annotation != 'Endothelial') %>% 
  group_by(Annotation, ebv_status) %>% 
  count() %>% 
  group_by(ebv_status) %>% 
  mutate(total_n = sum(n)) %>% 
  mutate(prop = n/total_n) %>% 
  ungroup() %>% 
  # group_by(Annotation, ebv_status) %>% 
  # summarise(celltype_mean_prop = mean(prop)) %>%
  # ungroup() %>% 
  group_by(Annotation) %>% 
  arrange(Annotation, ebv_status) %>% 
  mutate(diff = log2(prop/lag(prop))) %>% 
  drop_na() %>% 
  ggplot(aes(x = reorder(Annotation, diff), y = diff, fill = Annotation)) +
  geom_col() + 
  geom_rect(aes(xmin = 0, ymin = 0, xmax = Inf, ymax = Inf), fill = '#D55E00', alpha = 0.02) + 
  geom_rect(aes(xmin = 0, ymin = -Inf, xmax = Inf, ymax = 0), fill = '#009E73', alpha = 0.02) + 
  #annotate('text', x = 2, y = 1.5, label = 'EBV+', size = 12) +
  #annotate('text', x = 13, y = -1, label = 'EBV-', size = 12) +
  labs(x = 'Cell Type', y = 'Log2 Fold Enrichment', title = 'ROI Combined, Cell Type Enrichment (EBV+/EBV-)') + 
  scale_fill_manual('Cell Types', values = celltype_color, na.value = 'white', na.translate = FALSE) +
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5),
        legend.position = 'none',
        text = element_text(size = 12))

log2fc_p

```

## Fig4 F

```{r, eval = FALSE}

df <- INDEPTH_whole_core_df_scaled_norm %>% 
  # filter(str_detect(coreName, 'DFCI')) %>%
  filter(!str_detect(coreName, 'Tonsil')) %>% 
  filter(Annotation != 'Neutrophil') %>% 
  filter(Annotation != 'Other') %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  filter(Annotation %in% c('Tumor','M1-like', 'M2-like', 'DC', 'CD4 T', 'CD8 T', 'Treg', 'Endothelial'))

marker <- c('HLA1', 'HLA-DR', 'PD-L1')

pop_stat <- df %>%
  dplyr::select(coreName, Annotation, ebv_status, all_of(marker)) %>%
  pivot_longer(cols = -c('coreName', 'Annotation', 'ebv_status'), names_to = 'marker', values_to = 'value') %>%
  group_by(marker, coreName) %>%
  summarise(pop_mean = mean(value),
            pop_sd = sd(value)) %>% 
  group_by(marker) %>% 
  summarise(pop_mean = mean(pop_mean),
            pop_sd = mean(pop_sd)) %>% 
  ungroup()



stratified_stat <- df %>%
  dplyr::select(coreName, Annotation, ebv_status, all_of(marker)) %>%
  # filter(cluster != 'Other') %>%
  pivot_longer(cols = -c('coreName', 'Annotation', 'ebv_status'), names_to = 'marker', values_to = 'value') %>%
  group_by(marker, Annotation, ebv_status, coreName) %>%
  summarise(mu = mean(value)) %>% 
  group_by(marker, Annotation, ebv_status) %>% 
  summarise(mu = mean(mu)) %>% 
  ungroup()

bar_vec <- df %>%
  dplyr::select(coreName, Annotation, ebv_status, all_of(marker)) %>%
  # filter(cluster != 'Other')  %>%
  group_by(Annotation, ebv_status) %>%
  dplyr::count() %>%
  mutate(log_n = log10(n)) %>%
  arrange(ebv_status)

match_idx <- match(stratified_stat$marker, pop_stat$marker)

stratified_stat <- stratified_stat %>%
  mutate(z_score = (mu - pop_stat$pop_mean[match_idx])/pop_stat$pop_sd[match_idx]) %>%
  mutate(celltype_group1 = interaction(Annotation, ebv_status)) %>%
  arrange(celltype_group1)

heatmap_df <- stratified_stat %>%
  dplyr::select(marker, celltype_group1, z_score) %>%
  pivot_wider(id_cols = 'marker', names_from = 'celltype_group1', values_from = 'z_score')

heatmap_mat <- heatmap_df %>%
  column_to_rownames(var = 'marker') %>%
  as.matrix()



color_df <- as.data.frame(celltype_color) %>%
  rownames_to_column('celltype') %>%
  mutate(celltype = str_extract(celltype, '^[^.]+')) %>%
  distinct() %>%
  filter(celltype %in% c('Tumor','M1-like', 'M2-like', 'DC', 'CD4 T', 'CD8 T', 'Treg', 'Endothelial')) %>%
  arrange(celltype)

color_vec <- color_df$celltype_color

names(color_vec) <- color_df$celltype

column_ha <- HeatmapAnnotation(group1 = factor(str_extract(colnames(heatmap_mat), '(?<=\\.)\\w*'), levels = c('Positive', 'Negative')),
                               col = list(group1 = c('Positive' = '#D55E00', 'Negative' = '#009E73')),
                               
                               count = anno_barplot(bar_vec$n, border = FALSE, bar_width = 0.9,
                                                    gp = gpar(col = 'black', fill = rep(color_df$celltype_color, 2)),
                                                    height = unit(3, 'cm')),
                               show_annotation_name = FALSE,
                               annotation_legend_param = list(title = 'EBV Status'))

column_ha2 <- HeatmapAnnotation(celltype = rep(color_df$celltype,2),
                                col = list(celltype = color_vec), show_annotation_name = FALSE,
                                gp = gpar(col = 'black', lwd = 2),
                                annotation_legend_param = list(title = 'Cell Type'))

library(circlize)
col_fun = colorRamp2(c(-1, 0, 1), c('#4575b4', "white", '#d73027'))


h <- Heatmap(heatmap_mat, cluster_columns = FALSE, column_split = factor(str_extract(colnames(heatmap_mat), '(?<=\\.)\\w*'), levels = c('Positive', 'Negative')), 
             cluster_column_slices = FALSE, cluster_rows = FALSE, row_names_side = 'left', rect_gp = gpar(col = "black", lwd = 2), 
             top_annotation = column_ha, column_title = 'MHC1/2 PD1, Whole Core, Combined, norm, scaled', bottom_annotation = column_ha2, show_column_names = TRUE, 
             name = 'Z Score', heatmap_legend_param = list(border = 'black'), col = col_fun, row_order = marker)

draw(h)

```

## Fig4 F supp

```{r, eval = FALSE}

p <- INDEPTH_whole_core_df_scaled_norm %>% 
  # filter(str_detect(coreName, 'DFCI')) %>%
  filter(!str_detect(coreName, 'Tonsil')) %>% 
  filter(Annotation != 'Other') %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  filter(Annotation %in% c('Tumor','M1-like', 'M2-like', 'DC', 'CD4 T', 'CD8 T', 'Treg', 'Endothelial')) %>% 
  dplyr::select(coreName, Annotation, ebv_status, all_of(c('HLA1', 'HLA-DR', 'PD-L1'))) %>% 
  pivot_longer(cols = -c('coreName', 'Annotation', 'ebv_status'), names_to = 'marker', values_to = 'value') %>% 
  group_by(coreName, Annotation, ebv_status, marker) %>% 
  summarise(mu = mean(value)) %>% 
  ungroup() %>% 
  ggplot(aes(x = marker, y = mu, fill = ebv_status)) + 
  geom_boxplot(outlier.alpha = 0, position = position_dodge(1)) +
  geom_point(position = position_dodge(1)) + 
  facet_wrap(~Annotation) + 
  stat_compare_means(aes(group = ebv_status)) + 
  theme_bw() + 
  scale_fill_manual(values = c('#D55E00', '#009E73')) + 
  labs(x = 'Marker', y = 'Mean Marker Expression')

```



## Fig4 G

```{r, eval = FALSE}

T_df <- INDEPTH_whole_core_df_scaled_norm %>% 
  # filter(str_detect(coreName, 'DFCI')) %>%
  filter(!str_detect(coreName, 'Tonsil')) %>% 
  filter(Annotation != 'Neutrophil') %>% 
  filter(Annotation != 'Other') %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  filter(str_detect(Annotation, 'CD4') | str_detect(Annotation, 'CD8')) %>% 
  mutate(mem = ifelse(str_detect(Annotation, 'mem'), 1, 0),
         naive = ifelse(str_detect(Annotation, 'naive'), 1, 0)) %>% 
  mutate(Exhaustion_score = LAG3 + CD45RO - CD45RA + Tox - Ki67 - GZMB) %>% 
  mutate(Exhaustion_score = as.numeric(scale(Exhaustion_score)))

marker <- c('Exhaustion_score')

p <- T_df %>%
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>%
  dplyr::select(coreName, Annotation, ebv_status, all_of(marker)) %>%
  # filter(cluster != 'Other') %>%
  pivot_longer(cols = -c('coreName', 'Annotation', 'ebv_status'), names_to = 'marker', values_to = 'value') %>%
  group_by(marker, Annotation, ebv_status, coreName) %>%
  summarise(mu = mean(value)) %>% 
  ungroup() %>%
  mutate(marker_ebv = paste0(marker, '_', ebv_status)) %>% 
  mutate(marker_ebv = factor(marker_ebv, levels = c('Exhaustion_score_Positive', 'Exhaustion_score_Negative'))) %>% 
  ggplot(aes(x = marker_ebv, y = scale(mu), fill = ebv_status)) + 
  geom_violin(trim = FALSE) + 
  geom_boxplot(outlier.alpha = 0, width = 0.1) +
  geom_point(position = position_jitter(0.01), alpha = 1) + 
  stat_compare_means(comparisons = list(c('Exhaustion_score_Positive', 'Exhaustion_score_Negative')), method.args = list(alternative = 'greater')) + 
  scale_fill_manual(values = c('#D55E00', '#009E73')) +
  theme_bw() + 
  facet_wrap(~Annotation) + 
  labs(x = 'EBV Status', y = 'Exhaustion Score', fill = 'EBV Status')

p

```

## Fig4 I

The heatmap was generated in Python and therefore not included here.

```{r, eval = FALSE}

p <- niche_df %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  group_by(ebv_status, Niche, coreName) %>% 
  dplyr::count() %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  ggplot(aes(x = Niche, y = n, fill = ebv_status)) + 
  geom_violin(trim = TRUE, position = position_dodge(1), scale = 'width') + 
  geom_boxplot(outlier.alpha = 0, position = position_dodge(1), width = 0.2) +
  geom_point(position = position_dodge(1)) + 
  stat_compare_means(aes(group = ebv_status)) + 
  scale_fill_manual(values = c('#D55E00', '#009E73')) + 
  theme_bw() + 
  labs(y = 'Motif Count')

p

```

## Supp Fig6 A

```{r, eval = FALSE}

m1_m2_box <- INDEPTH_whole_core_df_scaled_norm %>% 
  # filter(!str_detect(coreName, 'Tonsil')) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'Tumor'), 'Tumor', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M1'), 'M1-like', Annotation)) %>% 
  mutate(Annotation = ifelse(str_detect(Annotation, 'M2'), 'M2-like', Annotation)) %>% 
  mutate(ebv_status = ifelse(coreName %in% c("DFCI_1.2", "DFCI_2.2", "DFCI_3.2",
                                             "DFCI_4.1", "DFCI_6.1", "DFCI_7.1", "DFCI_8.1", "DFCI_9.1", "DFCI_11.1",
                                             "DFCI_12.1", "Rochester_4", "Rochester_5","Rochester_7",
                                             "Rochester_9", "Rochester_10","Rochester_14", "Rochester_11", "Rochester_15",
                                             "Rochester_16", "Rochester_17", "Rochester_18", "Rochester_21", "Rochester_25"
  ), 'Positive', 'Negative')) %>% 
  # filter(Annotation != 'Neutrophil') %>%
  # filter(Annotation != 'Tumor') %>%
  # filter(Annotation != 'Other') %>%
  # filter(Annotation != 'Endothelial') %>% 
  filter(Annotation %in% c('M1-like', 'M2-like')) %>% 
  group_by(coreName, Annotation, ebv_status) %>% 
  count() %>% 
  group_by(coreName, ebv_status) %>% 
  mutate(total_n = sum(n)) %>% 
  mutate(prop = n/total_n) %>% 
  ungroup() %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  ggplot(aes(x = Annotation, y = prop, fill = ebv_status)) + 
  geom_violin(position = position_dodge(0.8)) + 
  geom_boxplot(outlier.alpha = 0, position = position_dodge(0.8), width = 0.1) +  
  geom_jitter(position = position_dodge(0.8)) +
  scale_fill_manual(values = c('#D55E00', '#009E73')) + 
  stat_compare_means(aes(group = ebv_status)) + 
  theme_bw() + 
  labs(fill = 'EBV Status', y = 'Proportion')

m1_m2_box

```


## Supp Fig6 D

```{r, eval = FALSE}

Niche_group_df <- niche_df %>% 
  mutate(group = case_when(Niche == 0 ~ 'Motif 1',
                           Niche %in% c(1,2,3,4) ~ 'Other')) %>%
  # mutate(group = case_when(Niche %in% c(0, 2) ~ 'Macrophage enriched',
  #                          Niche %in% c(1,3,4) ~ 'Other')) %>% 
  left_join(T_df %>%
               mutate(Annotation = ifelse(str_detect(Annotation, 'CD4'), 'CD4 T', Annotation)) %>% 
               mutate(Annotation = ifelse(str_detect(Annotation, 'CD8'), 'CD8 T', Annotation)) %>% 
               dplyr::select(coreName, cellLabel, Exhaustion_score), by = c('coreName', 'cellLabel'))

Niche_group_df <- Niche_group_df %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  mutate(new_group = paste0(group, '_', ebv_status))


mylist <- combn(unique(Niche_group_df$new_group), 2, simplify = FALSE)

p <- Niche_group_df %>% 
  mutate(new_group = factor(new_group, levels = c('Motif 1_Positive',  'Other_Positive', 'Motif 1_Negative',  'Other_Negative'))) %>%
  # mutate(new_group = factor(new_group, levels = c('Macrophage enriched_Positive', 'Other_Positive', 'Macrophage enriched_Negative', 'Other_Negative'))) %>%
  dplyr::select(coreName, new_group, Exhaustion_score, ebv_status) %>% 
  group_by(coreName, ebv_status, new_group) %>%
  summarise(mean_exhaustion = mean(Exhaustion_score)) %>% 
  ungroup() %>% 
  ggplot(aes(x = new_group, y = mean_exhaustion, fill = ebv_status)) + 
  geom_boxplot(outlier.alpha = 0) + 
  geom_point() + 
  stat_compare_means(comparisons = mylist) +
  scale_fill_manual(values = c('#D55E00', '#009E73')) + 
  theme_bw() 

p

```

```{r, eval = FALSE}

motif_heatmap_df <- Niche_group_df %>% 
  dplyr::select(coreName, group, Exhaustion_score, ebv_status)

marker <- c('Exhaustion_score')

pop_stat <- motif_heatmap_df %>%
  dplyr::select(coreName, group, ebv_status, all_of(marker)) %>%
  pivot_longer(cols = -c('coreName', 'group', 'ebv_status'), names_to = 'marker', values_to = 'value') %>%
  group_by(marker, coreName) %>%
  summarise(pop_mean = mean(value),
            pop_sd = sd(value)) %>% 
  group_by(marker) %>% 
  summarise(pop_mean = mean(pop_mean),
            pop_sd = mean(pop_sd)) %>% 
  ungroup()


stratified_stat <- motif_heatmap_df %>%
  dplyr::select(coreName, group, ebv_status, all_of(marker)) %>%
  # filter(cluster != 'Other') %>%
  pivot_longer(cols = -c('coreName', 'group', 'ebv_status'), names_to = 'marker', values_to = 'value') %>%
  group_by(marker, group, ebv_status, coreName) %>%
  summarise(mu = mean(value)) %>% 
  group_by(marker, group, ebv_status) %>% 
  summarise(mu = mean(mu)) %>% 
  ungroup()

bar_vec <- motif_heatmap_df %>%
  dplyr::select(coreName, group, ebv_status, all_of(marker)) %>%
  # filter(cluster != 'Other')  %>%
  group_by(group, ebv_status) %>%
  dplyr::count() %>%
  mutate(log_n = log10(n)) %>%
  arrange(ebv_status)

match_idx <- match(stratified_stat$marker, pop_stat$marker)

stratified_stat <- stratified_stat %>%
  mutate(z_score = (mu - pop_stat$pop_mean[match_idx])/pop_stat$pop_sd[match_idx]) %>%
  mutate(celltype_group1 = interaction(group, ebv_status)) %>%
  arrange(celltype_group1)

heatmap_df <- stratified_stat %>%
  dplyr::select(marker, celltype_group1, z_score) %>%
  pivot_wider(id_cols = 'marker', names_from = 'celltype_group1', values_from = 'z_score')

heatmap_mat <- heatmap_df %>%
  column_to_rownames(var = 'marker') %>%
  as.matrix()


column_ha <- HeatmapAnnotation(group1 = factor(str_extract(colnames(heatmap_mat), '(?<=\\.)\\w*'), levels = c('Positive', 'Negative')),
                               col = list(group1 = c('Positive' = '#D55E00', 'Negative' = '#009E73')),
                               
                               count = anno_barplot(bar_vec$n, border = FALSE, bar_width = 0.9,
                                                    gp = gpar(col = 'black', fill = rep('grey', 4)),
                                                    height = unit(3, 'cm')),
                               show_annotation_name = FALSE,
                               annotation_legend_param = list(title = 'EBV Status'))


library(circlize)
col_fun = colorRamp2(c(-0.5, 0, 0.5), c('#4575b4', "white", '#d73027'))


h <- Heatmap(heatmap_mat, cluster_columns = FALSE, column_split = factor(str_extract(colnames(heatmap_mat), '(?<=\\.)\\w*'), levels = c('Positive', 'Negative')), 
             cluster_column_slices = FALSE, cluster_rows = FALSE, row_names_side = 'left', rect_gp = gpar(col = "black", lwd = 2), 
             top_annotation = column_ha, column_title = 'Exhaustion score for Motif:EBV status', show_column_names = TRUE, 
             name = 'Z Score', heatmap_legend_param = list(border = 'black'), col = col_fun, row_order = marker)
draw(h)

```


## Supp Fig6 E

Here the proposed model is 


\[
\ln{E\left[Y_{i}\right]} = \beta_{0} + \beta_{1}I_{\mbox{EBV}} + \sum_{i=2}^{5}\beta_{i}I_{i} + \sum_{i=1}^{4}\gamma_{i}J_{\mbox{EBV}, i}
\]

where 

\[
\begin{aligned}
I_{\mbox{EBV}} &= \left\{
\begin{aligned}
&1, \mbox{EBV+}\\
&0, \mbox{EBV-}
\end{aligned}
\right., \\
\\
I_{i} &= \left\{
\begin{aligned}
&1, \mbox{Motif i}\\
&0, \mbox{Not Motif i}
\end{aligned}
\right. ,\\
\\
J_{EBV, i} &= \left\{
\begin{aligned}
&1, \mbox{EBV+, Motif i+1}\\
&0, \mbox{Not EBV+, Motif i+1}
\end{aligned}
\right. .
\end{aligned}
\]

Here the reference class is EBV negative and Motif 1.


```{r, eval = FALSE}

# fit regression

fit1 <- glm.nb(`M1-like` ~ ebv_status + Niche + ebv_status:Niche, data = niche_df)

fit2 <- glm.nb(`M2-like` ~ ebv_status + Niche + ebv_status:Niche, data = niche_df)

# summary

summary(fit1)

summary(fit2)

# Get CI

fit1_confint <- confint(fit1)

fit2_confint <- confint(fit2)

# Create prediction data frame
pred_df_m2 <- expand.grid(ebv_status = c("Negative", "Positive"),
                       Niche = factor(0:4))

# Create design matrix
X <- model.matrix(~ ebv_status + Niche + ebv_status:Niche, data = pred_df)

# Get coefficients and their CIs
coef_ci <- confint(fit2)
coef_est <- coef(fit2)

# Calculate predictions
pred_df_m2$expected <- as.numeric(exp(X %*% coef_est))
pred_df_m2$lower <- as.numeric(exp(X %*% coef_ci[,1]))  # lower CI bound
pred_df_m2$upper <- as.numeric(exp(X %*% coef_ci[,2]))  # upper CI bound


# Create prediction data frame
pred_df_m1 <- expand.grid(ebv_status = c("Negative", "Positive"),
                          Niche = factor(0:4))

# Create design matrix
X <- model.matrix(~ ebv_status + Niche + ebv_status:Niche, data = pred_df)

# Get coefficients and their CIs
coef_ci <- confint(fit1)
coef_est <- coef(fit1)

# Calculate predictions
pred_df_m1$expected <- as.numeric(exp(X %*% coef_est))
pred_df_m1$lower <- as.numeric(exp(X %*% coef_ci[,1]))  # lower CI bound
pred_df_m1$upper <- as.numeric(exp(X %*% coef_ci[,2]))  # upper CI bound



m2_plot <- pred_df_m2 %>%  
  mutate(sig = c(rep('sig', 9), 'non-sig')) %>% 
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  mutate(Motif = as.double(Niche)) %>% 
  mutate(Group = paste0('Motif ', Motif, ':', ebv_status)) %>% 
  mutate(Group = factor(Group, levels = c('Motif 1:Positive', 'Motif 2:Positive', 'Motif 3:Positive', 'Motif 4:Positive', 'Motif 5:Positive',
                                          'Motif 1:Negative', 'Motif 2:Negative', 'Motif 3:Negative', 'Motif 4:Negative', 'Motif 5:Negative'))) %>% 
  ggplot() +
  geom_errorbar(aes(x = Group, ymin = lower, ymax = upper, alpha = sig), width = 0.1) +
  geom_point(aes(x = Group, y = expected, color = ebv_status, alpha = sig), size = 3) +
  coord_flip() + 
  theme_bw() + 
  geom_hline(yintercept = pred_df_m2$expected[1], linetype = 'dashed') +
  labs(y = 'Expected M2 Count', x = 'Group') +
  scale_color_manual(values = c('#D55E00', '#009E73')) + 
  scale_alpha_manual(values = c('sig' = 1, 'non-sig' = 0.3)) + 
  labs(title = 'M2') + 
  theme(panel.grid.major.y = element_blank())


m1_plot <- pred_df_m1 %>%  
  mutate(sig = c('sig', 'sig', 'sig', 'sig', 'sig', 'non-sig', 'sig', 'non-sig', 'sig', 'non-sig')) %>%
  mutate(ebv_status = factor(ebv_status, levels = c('Positive', 'Negative'))) %>% 
  mutate(Motif = as.double(Niche)) %>% 
  mutate(Group = paste0('Motif ', Motif, ':', ebv_status)) %>% 
  mutate(Group = factor(Group, levels = c('Motif 1:Positive', 'Motif 2:Positive', 'Motif 3:Positive', 'Motif 4:Positive', 'Motif 5:Positive',
                                          'Motif 1:Negative', 'Motif 2:Negative', 'Motif 3:Negative', 'Motif 4:Negative', 'Motif 5:Negative'))) %>% 
  ggplot() +
  geom_errorbar(aes(x = Group, ymin = lower, ymax = upper, alpha = sig), width = 0.1) +
  geom_point(aes(x = Group, y = expected, color = ebv_status, alpha = sig), size = 3) +
  coord_flip() + 
  theme_bw() + 
  geom_hline(yintercept = pred_df_m1$expected[1], linetype = 'dashed') +
  labs(y = 'Expected M1 Count', x = 'Group') +
  scale_color_manual(values = c('#D55E00', '#009E73')) +
  scale_alpha_manual(values = c('sig' = 1, 'non-sig' = 0.3)) +
  labs(title = 'M1') + 
  theme(panel.grid.major.y = element_blank())

m1_plot

m2_plot

```



