---
layout: splash
title: "IN-situ DEtailed Phenotyping <br/> To High-resolution transcriptomics"

classes: wide
header:
  overlay_image: /assets/images/banner.png
  overlay_filter: 0.5
  # actions:
  #   - label: "üß™ Protocols"
  #     url: "/protocols/"
  #   - label: "üìì Notebooks"
  #     url: "/notebooks/"
  #   - label: "‚ÑπÔ∏è About"
  #     url: "/about/"
feature_row:
  - title: "Protocols"
    excerpt: "Step-by-step laboratory protocols."
    url: "/protocols/"
    btn_label: "View Protocols"
    btn_class: "btn--primary"
    # image_path: /assets/images/lab-logo.png
  - title: "Notebooks"
    excerpt: "Explore our Jupyter notebooks and analyses."
    url: "/notebooks/"
    btn_label: "See Notebooks"
    btn_class: "btn--primary"
    # image_path: /assets/images/lab-logo.png
  - title: "About the Lab"
    excerpt: "More about our team and research focus."
    url: "https://sizunjianglab.com/"
    btn_label: "Read More"
    btn_class: "btn--primary"
    # image_path: /assets/images/lab-logo.png
---



# üî¨ IN-DEPTH

Welcome to our documentation site.  
Use the menu to navigate to protocols and notebooks.

![IN-DEPTH Overview](/assets/images/overview.png){: .align-center}

The advent of spatial transcriptomics and spatial proteomics have enabled profound insights into tissue organization and provided systems-level understanding of diseases. However, both technologies remain largely independent, and emerging same slide spatial multi-omics approaches are generally limited in plex, spatial resolution, and analysis approaches. We introduce <ins>***IN-situ DEtailed Phenotyping To High-resolution transcriptomics (IN-DEPTH)***</ins>, a streamlined and resource-effective approach compatible with various spatial platforms. This iterative approach first entails single-cell spatial proteomics and rapid analysis to guide subsequent spatial transcriptomics capture on the same slide without loss in RNA signal. We also introduce k-bandlimited Spectral Graph Cross-Correlation (SGCC) to facilitate additional insights through integrative spatial multi-omics analysis. Application of IN-DEPTH and SGCC on lymphoid tissues demonstrated precise single-cell phenotyping and cell-type specific transcriptome capture, and accurately resolved the local and global transcriptome changes associated with the cellular organization of germinal centers. We then implemented IN-DEPTH and SGCC to dissect the tumor microenvironment (TME) of Epstein-Barr Virus (EBV)-positive and EBV-negative diffuse large B-cell lymphoma (DLBCL). Our results identified a key tumor-macrophage-CD4 T-cell immunomodulatory axis differently regulated between EBV-positive and EBV-negative DLBCL, and its central role in coordinating immune dysfunction and suppression. IN-DEPTH enables scalable, resource-efficient, and comprehensive spatial multi-omics dissection of tissues to advance clinically relevant discoveries.



{% include feature_row id="feature_row" %}
# Reference
